model/sharing/sharing.go

```diff
@@ -1056,206 +1056,337 @@ func CheckSharings(inst *instance.Instance) ([]map[string]interface{}, error) {
 				"id":         s.SID,
-				"type":       "not_enough_members",
-				"nb_members": len(s.Members),
+				"type":       "invalid_number_of_credentials",
+				"owner":      true,
+				"nb_members": len(s.Credentials),
 			})
-			return nil
+			return checks
+		}
+	} else {
+		if len(s.Credentials) != 1 {
+			checks = append(checks, map[string]interface{}{
+				"id":         s.SID,
+				"type":       "invalid_number_of_credentials",
+				"owner":      false,
+				"nb_members": len(s.Credentials),
+			})
+			return checks
 		}
 
-		for i, m := range s.Members {
-			if m.Status == MemberStatusRevoked && !s.Active {
-				continue
-			}
-			isFirst := i == 0
-			isOwner := m.Status == MemberStatusOwner
-			if isFirst != isOwner {
-				checks = append(checks, map[string]interface{}{
-					"id":     s.SID,
-					"type":   "invalid_member_status",
-					"member": i,
-					"status": m.Status,
-				})
-			}
+		if s.Credentials[0].InboundClientID == "" {
+			checks = append(checks, map[string]interface{}{
+				"id":    s.SID,
+				"type":  "missing_inbound_client_id",
+				"owner": false,
+			})
 		}
 
-		if !s.Active {
-			return nil
+		if !s.ReadOnly() && s.Members[0].Instance == "" {
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "missing_instance_for_member",
+				"member": 0,
+			})
+		}
+	}
+
+	return checks
+}
+
+func (s *Sharing) checkSharingTreesConsistency(inst *instance.Instance) (checks []map[string]interface{}) {
+	if !s.Owner || !s.Active || s.Initial {
+		return nil
+	}
+
+	rule := s.FirstFilesRule()
+	if rule == nil || !rule.HasSync() || !rule.FilesByID() {
+		// TODO: handle referenced_by type sharings?
+		return nil
+	}
+
+	isSharingReady := false
+	for _, m := range s.Members {
+		if m.Status == MemberStatusReady {
+			isSharingReady = true
+			break
 		}
+	}
+	if !isSharingReady {
+		return nil
+	}
 
-		if s.Owner {
-			for i, m := range s.Members {
-				if i == 0 || m.Status != MemberStatusReady {
+	ownersDocs, err := FindMatchingDocs(inst, *rule)
+	if err != nil {
+		checks = append(checks, map[string]interface{}{
+			"id":   s.SID,
+			"type": "missing_matching_docs_for_owner",
+			"err":  err,
+		})
+		return checks
+	}
+
+	for i, m := range s.Members {
+		member, err := instance.GetFromCouch(m.Instance)
+		if err != nil {
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "missing_instance_for_member",
+				"member": m.Instance,
+			})
+		}
+
+		memberDocs, err := FindMatchingDocs(member, *rule)
+		if err != nil {
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "missing_matching_docs_for_member",
+				"member": m.Instance,
+				"err":    err,
+			})
+			return checks
+		}
+
+		// Convert member doc ids to their owner's counterpart
+		memberKey := s.Credentials[i].XorKey
+		for _, doc := range memberDocs {
+			doc.SetID(XorID(doc.ID(), memberKey))
+		}
+
+		checkedMemberDocs := make([]bool, len(memberDocs))
```

Comment:
I would have made a `ownerDocsByID` map. Then, I can iterate on the memberDocs to find if they have a match in this map:

- if not, I can add a check error
- if yes, I can check that the two documents have the same revision, name, parent id (+checksum for file) and remove the doc from the map

At the end, I can list the doc remaining in the map and add an error for them (as they are only on the owner's Cozy).